### PR TITLE
Bacpop-130 Re-runnable projects

### PR DESCRIPTION
**Pull request by absternator**
This PR contains work to allow users to re-run their analysis after they have already run one. The following is done in the PR 
- allow users to add samples after run 
- allow them to re-rerun project 
- BE: add `hasRun` flag to keep track of if a sample has been run or not 

Testing: 
The best way to test this is to go ham with different combinations!!!!!!! Ensure all status on table are as expected
egs. 
i) new project add samples run analysis. then add more and re run analysis 
ii). try do same with some failing samples too 
iii). run a new project, add some samples then return to home page, back to project and then run again

### CODE REVIEW

File: app/server/src/controllers/indexController.ts [hunk 1]
```diff
@@ -16,7 +16,7 @@ export default (config) => {
                 const poppunkRequest = request.body as BeebopRunRequest;
                 const {projectHash, projectId, names, sketches} = poppunkRequest;
                 const {redis} = request.app.locals;
-                await userStore(redis).saveProjectHash(request, projectId, projectHash);
+                await userStore(redis).saveHashAndSampleHasRun(request, projectId, projectHash, names);
```
Comment (EmmaLRussell) [inline app/server/src/controllers/indexController.ts:19]: Am I right in thinking this is saving multiple sample names, so this method name maybe a little misleading?
```suggestion
                await userStore(redis).saveHashAndSamplesRun(request, projectId, projectHash, names);
```

**Review by EmmaLRussell — COMMENTED** [review]
This is great, works really nicely, and very intuitive. 

I found it a bit unexpected when new samples were moved to the top of the list when Run Analysis starts, having initially been added to the bottom. I can see that it makes sense, since the defaut pre-run behaviour is to append, and it makes sense to show the samples which are yet to run at the top - but I think it might be better to put new samples above any already-run samples as soon as you upload them so the order doesn't shift when you hit Run. 

One odd thing - here is the network view after a re-run (after removing some samples from previous run). The zoom looks wrong - some view settings retained from previous network view?
image (11)

**Comment by absternator** [comment]
> - but I think it might be better to put new samples above any already-run samples as soon as you upload them so the order doesn't shift when you hit Run.

1. the shifting is is because its re-ordering by alphabetical, not the only just run. Thus shift will still happen So I have left as if we add first may be a performance issue for 762 samples inserting first into the array.
2. good catch with the network tab zooming issue

File: app/client-v2/src/__tests__/components/ProjectView/ProjectFileUpload.spec.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    await waitFor(() => {
+      expect(store.onFilesUpload).toHaveBeenCalledWith([mockFile]);
+    });
+  });
+
+  it("should display running analysis when project is running", () => {
+    const testPinia = createTestingPinia();
+    const store = useProjectStore(testPinia);
+    // @ts-expect-error: Getter is read only
+    store.isRunning = true;
```
[the remark this answers is not in the document]
Comment (absternator) [reply app/client-v2/src/__tests__/components/ProjectView/ProjectFileUpload.spec.ts:83]: yeah initially i thought that too but seeems like it does error... and tried the util to wrap and looked at solutions(workarounds) out there to wrap and more ambigous honestly that just having the @ts-expect. also its constricted to this file only thus leaving

[the remark this answers is not in the document]
Comment (EmmaLRussell) [reply app/client-v2/src/__tests__/components/ProjectView/ProjectFileUpload.spec.ts:83]: Fair enough!

**Comment by EmmaLRussell** [comment]
> > * but I think it might be better to put new samples above any already-run samples as soon as you upload them so the order doesn't shift when you hit Run.
> 
>     1. the shifting is is because its re-ordering by alphabetical, not the only just run. Thus shift will still happen So I have left as if we add first may be a performance issue for 762 samples inserting first into the array.
> 
>     2. good catch with the network tab zooming issue

Oh, it's alphabetical! Ha, must just have been a coincidence that the last ones I added came first in the alphabet! OK, yeah, I think that's fine..

**Pull request merged**